content/blog/2017/2017-02-07-declarative-maven-project.adoc

```diff
@@ -0,0 +1,295 @@
+pipeline {
+    agent any
+    tools { // <1>
+        maven 'Maven 3.3.9' // <2>
+        jdk 'jdk8' // <3>
+    }
+    stages {
+        stage ('Initialize') {
+            steps {
+                sh '''
+                    echo "PATH = ${PATH}"
+                    echo "M2_HOME = ${M2_HOME}"
+                ''' // <4>
+            }
+        }
+
+        stage ('Build') {
+            steps {
+                echo 'This is a minimal pipeline.'
+            }
+        }
+    }
+}
+// Scripted Not Defined //
+----
+<1> `tools` section for adding tool settings.
+<2> Configure this pipeline to use the Maven version matching "Maven 3.3.9"
+(configured in "Managing Jenkins" -> "Global Tool Configuration").
+<3> Configure this pipeline to use the Maven version matching "jdk8"
+(configured in "Managing Jenkins" -> "Global Tool Configuration").
+<4> These will show the values of `PATH` and `M2_HOME` environment variables.
+
+
+When we run this updated Pipeline the same way we ran the first, we see that
+the Declarative Pipeline has added another stage called "Declarative: Tool
+Install": In the console output, we see that during this particular stage "Maven 3.3.9" gets installed,
+and the `PATH` and `M2_HOME` environment variables are set:
+
+image::/images/post-images/2017-02-07/junit-tools.png[Declarative Pipeline with Tools Section, role="center"]
+
+.Console Output
+[source]
+----
+{ ... truncated lines ... }
+[Pipeline] // stage
+[Pipeline] stage
+[Pipeline] { (Declarative: Tool Install)
+[Pipeline] tool
+Unpacking https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.3.9/apache-maven-3.3.9-bin.zip
+to /Users/bitwiseman/jenkins/agents/osx_mbp/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.3.9
+on osx_mbp
+[Pipeline] envVarsForTool
+[Pipeline] tool
+[Pipeline] envVarsForTool
+[Pipeline] }
+[Pipeline] // stage
+{ ... }
+PATH = /Library/Java/JavaVirtualMachines/jdk1.8.0_92.jdk/Contents/Home/bin:/Users/bitwiseman/jenkins/agents/osx_mbp/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.3.9/bin:...
+M2_HOME = /Users/bitwiseman/jenkins/agents/osx_mbp/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.3.9
+{ ... }
+----
+
+
+== Running a Maven Build
+
+Finally, running a Maven build is trivial.  The `tools` section already added
+Maven and JDK8 to the `PATH`, all we need to do is call `mvn install`.  It
+would be nice if I could split the build and the tests into separate stages,
+but Maven is famous for not liking when people do that, so I'll leave it alone
+for now.
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+    agent any
+    tools {
+        maven 'Maven 3.3.9'
+        jdk 'jdk8'
+    }
+    stages {
+        stage ('Initialize') {
+            steps {
+                sh '''
+                    echo "PATH = ${PATH}"
+                    echo "M2_HOME = ${M2_HOME}"
+                '''
+            }
+        }
+
+        stage ('Build') {
+            steps {
+                sh 'mvn install' // <1>
```

Comment:
Minor: Looking at your screenshots where the build stage only takes 135ms, maybe there should be `clean` in here as well, so that you don't have a build that just re-uses the already-generated results?